.github/workflows/check-changelog.yml

```diff
@@ -0,0 +1,49 @@
+name: Check Changelog
+# This check makes sure that the changelog is properly updated
+# when a PR introduces a change in a test file.
+# To bypass this check, label the PR with "No Changelog Needed".
+on:
+  pull_request:
+
+jobs:
+  check:
+    runs-on: ubuntu-latest
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'No Changelog Needed') == 0 }}
+    steps:
+      - name: Get PR number and milestone
+        run: |
+          echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+          echo "MILESTONE_TITLE=${{ github.event.pull_request.milestone.title }} >> $GITHUB_ENV
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: '0'
+      - name: Check the changelog
+        run: |
+          set -xe
+          changed_files=$(git diff --name-only origin/master)
+          # Changelog should be updated only if tests have been modified
+          if [[ "$changed_files" =~ tests ]]
+          then
+            all_changelogs=$(cat ./doc/whats_new/v*.rst)
+            if [[ "$all_changelogs" =~ :pr:\`$PR_NUMBER\` ]]
+            then
+              echo "Changelog has been updated."
+              # If the pull request is milestoned check the correspondent changelog
+              if exist -f ./doc/whats_new/v${MILESTONE_TITLE:0:4}.rst
```

Comment:
This doesn't seem to check that the PR is milestoned, but rather checks whether the corresponding changelog file exists.

Reply:
Yes.
1) If the PR is not milestoned the variable is empty and the file does not exist.
2) If the PR is milestoned for a r+2 the file does not exist.
3) If the PR is milestoned for a r+1 or older the file exist.

As long as the test is done in that order this will avoid two tests in the first item.